libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+	// Always resume
+	if (resume_thread (th, dbg->bits) == -1) {
+		ret = false;
+	}
+	CloseHandle (th);
+	return ret;
+}
+
+int w32_attach(RDebug *dbg, int pid) {
+	RIOW32Dbg *rio = dbg->user;
+	HANDLE ph = OpenProcess (PROCESS_ALL_ACCESS, FALSE, pid);
+	if (ph == (HANDLE)NULL) {
+		return -1;
+	}
+	if (!DebugActiveProcess (pid)) {
+		CloseHandle (ph);
+		return -1;
+	}
+	RList *threads = r_list_new ();
+	if (!threads) {
+		CloseHandle (ph);
+		return -1;
+	}
+	threads = w32_thread_list (dbg, pid, threads);
+	if (threads->length == 0) {
+		r_list_free (threads);
+		CloseHandle (ph);
+		return -1;
+	}
+	int tid = ((RDebugPid *)threads->head->data)->pid;
+	r_list_free (threads);
+	rio->ph = ph;
+	rio->debug = true;
+	return tid;
+}
+
+int w32_detach(RDebug *dbg, int pid) {
+	if (pid == -1) {
+		return false;
+	}
+	if (dbg->pid == pid) {
+		RIOW32Dbg *rio = dbg->user;
+		bool ret;
+		if (rio->debug) {
+			ret = DebugActiveProcessStop (pid);
+		}
+		CloseHandle (rio->ph);
+		rio->ph = NULL;
+		rio->debug = false;
+		return ret;
+	}
+	return false;
+}
+
+int w32_select(RDebug *dbg, int pid, int tid) {
+	RIOW32Dbg *rio = dbg->user;
+	if (rio->ph != (HANDLE)NULL) {
+		return true;
+	}
+	// hack to support w32dbg:// and attach://
+	DEBUG_EVENT de;
+	de.dwProcessId = pid;
+	de.dwThreadId = tid;
+	bool cont = true;
+	do {
+		if (!ContinueDebugEvent (de.dwProcessId, de.dwThreadId, DBG_CONTINUE)) {
+			break;
+		}
+		if (!WaitForDebugEvent (&de, 1000)) {
+			break;
+		}
+		switch (de.dwDebugEventCode) {
+		//case CREATE_PROCESS_DEBUG_EVENT:
+		//case CREATE_THREAD_DEBUG_EVENT:
+		case LOAD_DLL_DEBUG_EVENT:
+			{
+				HANDLE hf = de.u.LoadDll.hFile;
+				if (hf && hf != INVALID_HANDLE_VALUE) {
+					CloseHandle (hf);
+				}
+			} break;
+		case EXCEPTION_DEBUG_EVENT:
+			// TODO: check for the type of exception?
+			cont = false;
+			break;
+		default:
+			eprintf ("Unhandled debug event %d\n", de.dwDebugEventCode);
+			break;
+		}
+	} while (cont);
+	rio->ph = OpenProcess (PROCESS_ALL_ACCESS, FALSE, pid);
+	if (rio->ph == (HANDLE)NULL) {
+		return false;
+	}
+	rio->debug = true;
+	return true;
+}
+
+int w32_kill(RDebug *dbg, int pid, int tid, int sig) {
+	if (sig == 0) {
+		return true;
+	}
+	RIOW32Dbg *rio = dbg->user;
+	if (rio->debug) {
+		DebugActiveProcessStop (pid);
+	}
+	bool ret = false;
+	if (TerminateProcess (rio->ph, 1)) {
+		if (WaitForSingleObject (rio->ph, 1000) != WAIT_OBJECT_0) {
+			r_sys_perror ("w32_kill/WaitForSingleObject");
+		} else {
+			ret = true;
+		}
+	}
+	CloseHandle (rio->ph);
+	rio->ph = NULL;
+	return ret;
+}
+
+void w32_break_process_wrapper(void *d) {
+	w32_break_process (d);
+}
+
+void w32_break_process(RDebug *dbg) {
+	RIOW32Dbg *rio = dbg->user;
+	if (!DebugBreakProcess (rio->ph)) {
+		r_sys_perror ("w32_break_process/DebugBreakProcess");
+	}
+}
+
+static char *get_w32_excep_name(unsigned long code) {
+	char *desc;
+	switch (code) {
+	/* fatal exceptions */
+	case EXCEPTION_ACCESS_VIOLATION:
+		desc = "access violation";
```

Comment:
i would go for just return "access violation", so you can save the local var and all the breaks.

also what is returned is `const char *`, not `char*`

Reply:
That code isn't mine actually but it was just 'migrated' from w32.c